### PR TITLE
port terminal-name

### PR DESCRIPTION
**Pull request by brotzeit**

### CODE REVIEW

**Comment by shanavas786** [comment]
permission of some files has been changed from `100644 → 100755`

File: src/terminal.c [hunk 1]
```diff
@@ -460,21 +430,6 @@ DEFUN ("terminal-list", Fterminal_list, Sterminal_list, 0, 0, 0,
...
-       doc: /* Return the name of the terminal device TERMINAL.
-It is not guaranteed that the returned value is unique among opened devices.
-
-TERMINAL may be a terminal object, a frame, or nil (meaning the
-selected frame's terminal). */)
-  (Lisp_Object terminal)
-{
-  struct terminal *t = decode_live_terminal (terminal);
-
-  return t->name ? build_string (t->name) : Qnil;
```
Comment (Wilfred) [inline src/terminal.c]: I can't see the null check of the name in the Rust. Have I missed it, or do we need to add it?

Comment (brotzeit) [reply src/terminal.c]: This is already checked in `decode_terminal` with `return t && t->name ? t : NULL;`. I used the method `is_live` in rust.

File: rust_src/src/terminal.rs [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    let t;
+    if let Some(mut term) = terminal.as_terminal() {
+        t = term.as_mut();
+    } else if let Some(frame) = terminal.as_frame() {
+        t = frame.terminal;
+    } else {
+        t = ptr::null_mut();
+    }
+
+    if let Some(term_ref) = LispTerminalRef::from_ptr(t as *mut c_void) {
```
Comment (shanavas786) [inline rust_src/src/terminal.rs]: I guess this causes segfault
We should check for null before transmuting to `LispTerminalRef`

Comment (shanavas786) [reply rust_src/src/terminal.rs]: We could assert that pointer is not null in `from_prt` methods too

Comment (brotzeit) [reply rust_src/src/terminal.rs]: Oh, I thought this is handled by `as_ref`.

File: rust_src/src/terminal.rs [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+/// Return the terminal object specified by TERMINAL.  TERMINAL may
+/// be a terminal object, a frame, or nil for the terminal device of
+/// the current frame.  If TERMINAL is neither from the above or the
+/// resulting terminal object is deleted, return NULL.
+#[no_mangle]
+pub extern "C" fn decode_terminal(mut terminal: LispObject) -> *mut Lisp_Terminal {
+    if terminal.is_nil() {
+        terminal = selected_frame();
+    }
+
```
Comment (shaleh) [inline rust_src/src/terminal.rs]: Why not assign to `t` here?

File: rust_src/src/terminal.rs (hunk 2 above)
Comment (shaleh) [reply rust_src/src/terminal.rs]: `ptr.as_ref` handles whether the pre is null. That is why the return value is an Option type.

Comment (shaleh) [reply rust_src/src/terminal.rs]: https://doc.rust-lang.org/std/primitive.pointer.html#method.as_ref

**Comment by brotzeit** [comment]
Why is this suddenly fixed ?

**Comment by shaleh** [comment]
@brotzeit I think I have this one resolved. I updated with the style we have been using lately. `LispLiveTerminal` will be used for many of the soon to be ported terminal functions.

**Comment by brotzeit** [comment]
Ok. I can't approve =)

**Comment by shaleh** [comment]
I can but I shouldn't.

**Pull request merged**